In [2]:
import pandas as pd

Import all the files

In [3]:
visits = pd.read_csv('visits.csv',
                     parse_dates=[1])
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
                   
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])

Step 1: Inspect the DataFrames using `print` and `head`

In [4]:
print(visits.head(5))
print(cart.head(5))
print(checkout.head(5))
print(purchase.head(5))

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 

Step 2: Left merging visits and cart

In [10]:
visits_cart = pd.merge(visits, cart, how='left')
visits_cart

user_id          visit_time  \
0     943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1     0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2     6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3     6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4     a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
...                                    ...                 ...   
1995  33913ac2-03da-45ae-8fc3-fea39df827c6 2017-03-25 03:29:00   
1996  4f850132-b99d-4623-80e6-6e61d003577e 2017-01-08 09:57:00   
1997  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70 2017-09-07 12:56:00   
1998  b01bffa7-63ba-4cd3-9d93-eb1477c23831 2017-07-20 04:37:00   
1999  0336ca81-8d68-443f-9248-ac0b8ad147d5 2017-11-15 10:11:00   

               cart_time  
0                    NaT  
1    2017-01-26 14:44:00  
2    2017-08-20 08:31:00  
3                    NaT  
4                    NaT  
...                  ...  
1995                 NaT  
1996                 NaT  
1997                 NaT  
1998                 NaT  
1999                 NaT  

[2000 rows x 3 columns]

Step 3: How long is `visits_cart`?

In [16]:
visits_cart.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2000 entries, 0 to 1999
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   user_id     2000 non-null   object        
 1   visit_time  2000 non-null   datetime64[ns]
 2   cart_time   348 non-null    datetime64[ns]
dtypes: datetime64[ns](2), object(1)
memory usage: 62.5+ KB


visits_cart is 2000 entries long, with 348 non-null entries

Step 4: How many timestamps are null for `cart_time`?

In [13]:
timestamps = visits_cart.cart_time.count()
timestamps

348

Step 5: What percentage only visited?

In [20]:
visited = len(visits)
only_visited = (visited - timestamps) / visited
only_visited

0.826

Step 6: What percentage placed a t-shirt in their cart but did not checkout?

In [25]:
cart_checkout = pd.merge(cart, checkout, how='left')
carts_made = len(cart)
checkouts = cart_checkout.checkout_time.count()
no_checkout_percent  = (carts_made - checkouts) / carts_made
no_checkout_percent

0.3505747126436782

Step 7: Merge it all together

In [28]:
all_data = visits.merge(cart, how='left').merge(checkout, how='left').merge(purchase, how='left')
all_data.head(15)

user_id          visit_time  \
0   943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1   0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2   6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3   6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4   a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
5   7bc62113-d51d-4e6b-85e0-1cf29e3de74a 2017-06-03 20:05:00   
6   6f22cd44-bc63-4449-a9af-2651859da2f9 2017-03-18 23:09:00   
7   49a6399b-c095-4e42-92eb-af90abe2bb6a 2017-01-10 12:09:00   
8   653c095d-fd74-40dd-ae3b-5a69a2613629 2017-06-27 19:21:00   
9   3d9fdac1-aae3-4a01-b448-934d12047468 2017-09-24 06:43:00   
10  1955acc9-5228-4277-beed-fd23f0f355b3 2017-04-20 23:01:00   
11  a627ba57-b34d-4166-8b78-1ff3f46f5478 2017-08-11 13:41:00   
12  10d5a3e0-1d72-4bea-9670-f0f0234d6e48 2017-02-26 10:09:00   
13  0c1fa5fe-27c1-4e1f-b353-26a428da2832 2017-07-04 09:56:00   
14  486480e2-98c3-4d51-8f4b-b1c07228ce84 2017-01-27 16:34:00   

             cart_time       checkout_time       purchase_time  
0                  NaT                 NaT                 NaT  
1  2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2  2017-08-20 08:31:00                 NaT                 NaT  
3                  NaT                 NaT                 NaT  
4                  NaT                 NaT                 NaT  
5                  NaT                 NaT                 NaT  
6                  NaT                 NaT                 NaT  
7                  NaT                 NaT                 NaT  
8                  NaT                 NaT                 NaT  
9                  NaT                 NaT                 NaT  
10                 NaT                 NaT                 NaT  
11                 NaT                 NaT                 NaT  
12                 NaT                 NaT                 NaT  
13                 NaT                 NaT                 NaT  
14 2017-01-27 16:44:00 2017-01-27 17:10:00 2017-01-27 17:12:00

Step 8: % of users who got to checkout but did not purchase

In [29]:
make_checkout = all_data.checkout_time.count()
make_purchase = all_data.purchase_time.count()
no_purchase = (make_checkout - make_purchase) / make_checkout
no_purchase

0.24550898203592814

Step 9: check each part of the funnel, let's print all 3 of them again

In [31]:
print(f'Visit but dont add shirt: {only_visited*100}%\nAdd\'s to cart but dont checkout: {no_checkout_percent * 100}%\nGoes to checkout but dont purchase: {no_purchase* 100}%\n')

Visit but dont add shirt: 82.6%
Add's to cart but dont checkout: 35.05747126436782%
Goes to checkout but dont purchase: 24.550898203592812%



*The weakest part of the funnel is clearly getting a person who visited the site to add a tshirt to their cart. Once they've added a t-shirt to their cart it is fairly likely they end up purchasing it. A suggestion could be to make the add-to-cart button more prominent on the front page.*


Step 10: adding new column

In [34]:
all_data['time_to_purchase'] = all_data.purchase_time - all_data.visit_time

Step 11: examine the results

In [35]:
all_data

user_id          visit_time  \
0     943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1     0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2     6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3     6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4     a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
...                                    ...                 ...   
2103  33913ac2-03da-45ae-8fc3-fea39df827c6 2017-03-25 03:29:00   
2104  4f850132-b99d-4623-80e6-6e61d003577e 2017-01-08 09:57:00   
2105  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70 2017-09-07 12:56:00   
2106  b01bffa7-63ba-4cd3-9d93-eb1477c23831 2017-07-20 04:37:00   
2107  0336ca81-8d68-443f-9248-ac0b8ad147d5 2017-11-15 10:11:00   

               cart_time       checkout_time       purchase_time  \
0                    NaT                 NaT                 NaT   
1    2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00   
2    2017-08-20 08:31:00                 NaT                 NaT   
3                    NaT                 NaT                 NaT   
4                    NaT                 NaT                 NaT   
...                  ...                 ...                 ...   
2103                 NaT                 NaT                 NaT   
2104                 NaT                 NaT                 NaT   
2105                 NaT                 NaT                 NaT   
2106                 NaT                 NaT                 NaT   
2107                 NaT                 NaT                 NaT   

     time_to_purchase  
0                 NaT  
1     0 days 00:44:00  
2                 NaT  
3                 NaT  
4                 NaT  
...               ...  
2103              NaT  
2104              NaT  
2105              NaT  
2106              NaT  
2107              NaT  

[2108 rows x 6 columns]

Step 12: average time to purchase

In [36]:
average_purchase_time = all_data.time_to_purchase.mean()
average_purchase_time

Timedelta('0 days 00:43:12.380952380')